File: LetterCombinationsOfAPhoneNumber.py
"""
17. Letter Combinations of a Phone Number
Given a digit string, return all possible letter combinations
that the number could represent.
"""
# O(n * 4^n)
# O(n)
class Solution(object):
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        if not digits:
            return []
        lookup = {"2":"abc", "3":"def", "4":"ghi", "5":"jkl", "6":"mno", "7":"pqrs", "8":"tuv", "9":"wxyz"}
        res, cur = [], ""
        self.recur(res, cur, digits, lookup)
        return res
        
    def recur(self, res, cur, digits, lookup):
        if not digits:
            res.append(cur)
            return
        if digits[0] not in lookup:
            self.recur(res, cur, digits[1:], lookup)
            return
        for i in range(len(lookup[digits[0]])):
            cur += lookup[digits[0]][i]
            self.recur(res, cur, digits[1:], lookup)
            cur = cur[0:len(cur) - 1]

if __name__ == "__main__":
    digits = "1*3#9"
    print(Solution().letterCombinations(digits))

File: Java/Letter Combinations of a Phone Number.java
import java.util.*;
public class letterCombinations {
	
	private static final Map<Character, String> lookup = new HashMap<Character, String>() {
        {
            put('2', "abc");
            put('3', "def");
            put('4', "ghi");
            put('5', "jkl");
            put('6', "mno");
            put('7', "pqrs");
            put('8', "tuv");
            put('9', "wxyz");
        }
    };  
            
	public static List<String> letterCombinations(String digits) {
	    List<String> res = new ArrayList<String>();
	    StringBuffer cur = new StringBuffer();
	    if(digits.length() != 0) {
	        String tem = lookup.get(digits.charAt(0));
	        for(int i = 0; i < tem.length(); i++) {
	            cur.append(tem.charAt(i));
	            recur(res, cur, 1, digits);
	            cur.deleteCharAt(cur.length() - 1);
	        }
	    }
	    return res;
	}
	
	public static void recur(List<String> res, StringBuffer cur, int start, String digits) {
	    if(start >= digits.length()) {
	        res.add(cur.toString());
	    } else {
	        String tem = lookup.get(digits.charAt(start));
	        for(int i = 0; i < tem.length(); i++) {
	            cur.append(tem.charAt(i));
	            recur(res, cur, start + 1, digits);
	            cur.deleteCharAt(cur.length() - 1);
	        }
	    }
	}

	public static void main(String[] args) {
		String digits = "23";
		List<String> res = letterCombinations(digits);
		System.out.println(res);

	}

}
